consensus/src/quorum_store/batch_requester.rs

```diff
@@ -111,39 +124,47 @@ impl<T: QuorumStoreSender + 'static> BatchRequester<T> {
         signers: Vec<PeerId>,
         ret_tx: oneshot::Sender<Result<Vec<SignedTransaction>, Error>>,
     ) {
-        let mut request_state = BatchRequesterState::new(signers, ret_tx);
+        let mut request_state = BatchRequesterState::new(signers, ret_tx, self.retry_limit);
         let network_sender = self.network_sender.clone();
         let request_num_peers = self.request_num_peers;
         let my_peer_id = self.my_peer_id;
         let epoch = self.epoch;
-        let timeout = Duration::from_millis(self.request_timeout_ms as u64);
+        let retry_timeout = Duration::from_millis(self.retry_timeout_ms as u64);
+        let rpc_timeout = Duration::from_millis(self.rpc_timeout_ms as u64);
 
         tokio::spawn(async move {
-            while let Some(request_peers) = request_state.next_request_peers(request_num_peers) {
+            monitor!("batch_request", {
+                let mut interval = time::interval(retry_timeout);
                 let mut futures = FuturesUnordered::new();
-                trace!("QS: requesting from {:?}", request_peers);
                 let request = BatchRequest::new(my_peer_id, epoch, digest);
-                for peer in request_peers {
-                    counters::SENT_BATCH_REQUEST_COUNT.inc();
-                    futures.push(network_sender.request_batch(request.clone(), peer, timeout));
-                }
-                while let Some(response) = futures.next().await {
-                    match response {
-                        Ok(batch) => {
-                            counters::RECEIVED_BATCH_RESPONSE_COUNT.inc();
-                            let digest = *batch.digest();
-                            let payload = batch.into_transactions();
-                            request_state.serve_request(digest, Some(payload));
-                            return;
-                        },
-                        Err(e) => {
-                            error!("Batch request failed: {}", e);
+                loop {
+                    tokio::select! {
+                        _ = interval.tick() => {
+                            // send batch request to a set of peers of size request_num_peers
+                            if let Some(request_peers) = request_state.next_request_peers(request_num_peers) {
+                                for peer in request_peers {
+                                    futures.push(network_sender.request_batch(request.clone(), peer, rpc_timeout));
+                                }
+                            } else {
+                                // retry_limit exceeds, fail to get batch from peers
+                                break;
```

Comment:
Hm, so if we break here, the last set of peers gets only 1 second to receive RPC responses? Instead of a break immediately, is the behavior we want to wait until the futures are drained?

Reply:
Good point! Resolved.